### PR TITLE
RUBY-2007 Clean up SSLSocket usage

### PR DESCRIPTION
**Pull request by egiurleo**
* Improve URI parsing
* Do not use the driver-specific implementation of SSL sockets
* Make sure to set nodelay option, perform SNI, and set a timeout
* Improve error handling

### CODE REVIEW

File: lib/mongo/crypt/encryption_io.rb [hunk 1]
```diff
@@ -95,29 +95,58 @@ def mark_command(cmd)
...
         # There is no specific timeout written in the spec. See SPEC-1394
         # for a discussion and updates on what this timeout should be.
         socket_timeout = 10
+        default_port = 443
 
-        host, port = endpoint.split(':')
-        port ||= 443
+        unless endpoint.start_with?('https://') || endpoint.start_with?('//')
+          endpoint = "//#{endpoint}"
+        end
```
Comment (egiurleo) [inline lib/mongo/crypt/encryption_io.rb]: My goal here is to make sure that URIs are correctly parsed in as many situations as possible. By default, LMC returns the KMS endpoint in the format `kms.us-east-1.amazonaws.com`, which confuses Ruby's URI parser because there is no scheme. Is there a better or more reliable way to handle this case?

File: lib/mongo/crypt/binding.rb [hunk 2]
```diff
@@ -793,7 +793,7 @@ def self.kms_ctx_message(kms_context)
           mongocrypt_kms_ctx_message(kms_context.kms_ctx_p, binary.ref)
         end
 
-        return binary.to_string
+        return binary.to_s
```
Comment (p-mongo) [inline lib/mongo/crypt/binding.rb]: Can this fix be in a separate PR please with the ticket reference to the to_string -> to_s change?

File: lib/mongo/crypt/encryption_io.rb (hunk 1 above)
Comment (p-mongo) [reply lib/mongo/crypt/encryption_io.rb]: https://github.com/mongodb/libmongocrypt/blob/master/integrating.md#state-mongocrypt_ctx_need_kms says:

> Create/reuse a TLS socket connected to the endpoint indicated by mongocrypt_kms_ctx_endpoint. The endpoint string is a host name with optional port number separated by a colon

So, this is actually supposed to be a URI, but just the host name, right?

Comment (egiurleo) [reply lib/mongo/crypt/encryption_io.rb]: Okay. In that case, do you think it's fair for me to just split the string at the colon and then raise an exception if the result doesn't have two parts? I can also specify in the documentation that any user-specified endpoint should take this same format.

Comment (p-mongo) [reply lib/mongo/crypt/encryption_io.rb]: Yes per our discussion just now.

File: lib/mongo/crypt/encryption_io.rb [hunk 3]
```diff
@@ -95,29 +95,58 @@ def mark_command(cmd)
       #   the endpoint at which to establish a TLS connection and the message
       #   to send on that connection.
       def feed_kms(kms_context)
-        endpoint = kms_context.endpoint
-        message = kms_context.message
+        with_ssl_socket(kms_context.endpoint) do |ssl_socket|
+          ssl_socket.puts(kms_context.message)
```
Comment (p-mongo) [inline lib/mongo/crypt/encryption_io.rb]: s/puts/write/, puts adds another newline

Comment (p-mongo) [reply lib/mongo/crypt/encryption_io.rb]: The write call should have a timeout.

File: lib/mongo/crypt/encryption_io.rb [hunk 4]
```diff
@@ -95,29 +95,58 @@ def mark_command(cmd)
...
 
+          while kms_context.bytes_needed > 0 do
+            bytes = ssl_socket.sysread(kms_context.bytes_needed)
```
Comment (p-mongo) [inline lib/mongo/crypt/encryption_io.rb]: This read call should also have a timeout.

File: lib/mongo/crypt/encryption_io.rb [hunk 5]
```diff
@@ -95,29 +95,58 @@ def mark_command(cmd)
       #   the endpoint at which to establish a TLS connection and the message
       #   to send on that connection.
       def feed_kms(kms_context)
-        endpoint = kms_context.endpoint
-        message = kms_context.message
+        with_ssl_socket(kms_context.endpoint) do |ssl_socket|
+          ssl_socket.puts(kms_context.message)
 
+          while kms_context.bytes_needed > 0 do
```
Comment (p-mongo) [inline lib/mongo/crypt/encryption_io.rb]: Please save bytes needed into a variable.

File: lib/mongo/crypt/encryption_io.rb [hunk 6]
```diff
@@ -95,29 +101,68 @@ def mark_command(cmd)
...
 
-          kms_context.feed(bytes.pack('C*'))
-          num_bytes_needed = kms_context.bytes_needed
+          yield(ssl_socket)
+        rescue => e
+          raise Error::KmsError, "Error decrypting data key. #{e.class}: #{e.message}"
+        ensure
+          # If there is an error during socket creation, the
+          # ssl_socket object won't exist in this scope
+          ssl_socket.sysclose if ssl_socket
```
Comment (p-mongo) [inline lib/mongo/crypt/encryption_io.rb]: This close could conceivably also fail, especially since it is an ssl socket and close may perform i/o. Either putting the close call under the rescue raising KmsError above or adding `rescue nil` should be reasonable measures to deal with this. I don't know if the close call should also be guarded with a timeout; I suppose it won't hurt if you wish to do this.
